client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQRealPushConsumerImpl.java

```diff
@@ -0,0 +1,1305 @@
+        long computeAccTotal = this.computeAccumulationTotal();
+        long adjustThreadPoolNumsThreshold = this.defaultMQPushConsumer.getAdjustThreadPoolNumsThreshold();
+
+        long incThreshold = (long) (adjustThreadPoolNumsThreshold * 1.0);
+
+        long decThreshold = (long) (adjustThreadPoolNumsThreshold * 0.8);
+
+        if (computeAccTotal >= incThreshold) {
+            this.consumeMessageService.incCorePoolSize();
+        }
+
+        if (computeAccTotal < decThreshold) {
+            this.consumeMessageService.decCorePoolSize();
+        }
+    }
+
+    private long computeAccumulationTotal() {
+        long msgAccTotal = 0;
+        ConcurrentMap<MessageQueue, ProcessQueue> processQueueTable = this.rebalanceImpl.getProcessQueueTable();
+        Iterator<Entry<MessageQueue, ProcessQueue>> it = processQueueTable.entrySet().iterator();
+        while (it.hasNext()) {
+            Entry<MessageQueue, ProcessQueue> next = it.next();
+            ProcessQueue value = next.getValue();
+            msgAccTotal += value.getMsgAccCnt();
+        }
+
+        return msgAccTotal;
+    }
+
+    public List<QueueTimeSpan> queryConsumeTimeSpan(final String topic)
+        throws RemotingException, MQClientException, InterruptedException, MQBrokerException {
+        List<QueueTimeSpan> queueTimeSpan = new ArrayList<QueueTimeSpan>();
+        TopicRouteData routeData = this.mQClientFactory.getMQClientAPIImpl().getTopicRouteInfoFromNameServer(topic, 3000);
+        for (BrokerData brokerData : routeData.getBrokerDatas()) {
+            String addr = brokerData.selectBrokerAddr();
+            queueTimeSpan.addAll(this.mQClientFactory.getMQClientAPIImpl().queryConsumeTimeSpan(addr, topic, groupName(), 3000));
+        }
+
+        return queueTimeSpan;
+    }
+
+    public ConsumeMessageService getConsumeMessageService() {
+        return consumeMessageService;
+    }
+
+    public void setConsumeMessageService(ConsumeMessageService consumeMessageService) {
+        this.consumeMessageService = consumeMessageService;
+
+    }
+
+    private void tryToFindSnodePublishInfo() {
+        this.mQClientFactory.updateSnodeInfoFromNameServer();
+    }
+
+    public boolean processPushMessage(final MessageExt msg,
+        final String consumerGroup,
+        final String topic,
+        final String brokerName,
+        final int queueID,
+        final long offset) {
+        String localOffsetKey = genLocalOffsetKey(consumerGroup, topic, brokerName, queueID);
+        AtomicLong localOffset = this.localConsumerOffset.get(localOffsetKey);
+        if (localOffset == null) {
+            log.info("Current Local offset have not set, initiallized to -1.");
+            this.localConsumerOffset.put(localOffsetKey, new AtomicLong(-1));
+            return false;
+        }
+        AtomicBoolean pullStop = this.pullStopped.get(localOffsetKey);
+        if (pullStop == null) {
+            this.pullStopped.put(localOffsetKey, new AtomicBoolean(false));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **1227, 1228, and 1229**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread's `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).